### PR TITLE
New preferences page layout

### PR DESCRIPTION
**Pull request by roryabraham**
HOLD on https://github.com/Expensify/App/pull/20663

### Details
This PR was split off of https://github.com/Expensify/App/pull/15462. It implements a new layout in the `Settings` -> `Preferences` page.

### Fixed Issues
$ (partial) https://github.com/Expensify/App/issues/12261

### Tests
1. Open the preferences page.
1. Make sure it matches the screenshots.
1. Test the UI for regressions.

- [x] Verify that no errors appear in the JS console

### Offline tests
Same as tests.

### QA Steps
Same as tests.

- [ ] Verify that no errors appear in the JS console

### PR Author Checklist
- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://github.com/Expensify/App/assets/47436092/770c5beb-6904-4624-bf17-99340ca79f31

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://github.com/Expensify/App/assets/47436092/7afe5e17-9966-4e7f-9df2-1b10e7ce4f79

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://github.com/Expensify/App/assets/47436092/83b059d7-15e0-401b-896c-d74d6294b2ec

</details>

<details>
<summary>Desktop</summary>

https://github.com/Expensify/App/assets/47436092/80ffaebe-37db-4f1f-9ac4-d6d9c686ebbc

</details>

<details>
<summary>iOS</summary>

https://github.com/Expensify/App/assets/47436092/35663ad2-cdb1-484d-bee9-76273fc59d3b

</details>

<details>
<summary>Android</summary>

https://github.com/Expensify/App/assets/47436092/5aa687d2-935a-4df3-85d1-e99edcaac38c

</details>

### CODE REVIEW

**Comment by roryabraham** [comment]
This PR is branched off of https://github.com/Expensify/App/pull/20663, so the diff will be smaller once that's merged.

**Comment by roryabraham** [comment]
@allroundexperts is going to review this

File: patches/react-native-web-lottie+1.4.4.patch
Comment (roryabraham) [inline patches/react-native-web-lottie+1.4.4.patch:18]: I found this seems to yield better framerates and also fixes a display but on iOS Safari with different animation (not yet included in the app)

File: src/components/HeaderWithBackButton/index.js
Comment (roryabraham) [inline src/components/HeaderWithBackButton/index.js:1]: IDK why this shows as deleted + new file instead of renamed 😏

**Comment by roryabraham** [comment]
@allroundexperts refresh the page maybe? GitHub was having issues but I posted all the screenshots before putting this in for review. I can still see them in the issue description

**Comment by roryabraham** [comment]
Here's the mockup that @shawnborton provided: 

image

**Comment by roryabraham** [comment]
Good catch @allroundexperts – investigating now

**Comment by shawnborton** [comment]
Otherwise the screenshots look pretty good to me!

**Comment by roryabraham** [comment]
Just updated the scrolling behavior (among a few other small fixes). I am aware of a minor issue on iOS and iOS Safari where the animation can overscroll beyond its blue header:

https://github.com/Expensify/App/assets/47436092/e4314b6f-a7ae-48b6-b050-ade563f8b244

I fixed the same issue on the bottom (where before you could overscroll on the bottom and see some blue peeking out). However, fixing the same thing on the top caused scroll jankiness b/c the "spacer view" that we need to expand to keep blue beneath the animation must start w/ height 0, and then we can't possibly re-render it in JS fast enough to keep up with native scroll events.

Anyways, I don't think it's really a problem so this should be ready for another review.

**Comment by shawnborton** [comment]
Looks like we get some weirdness with the screen height is not very tall:
image


That blue BG part should just remain the same height.

And of course this is an extreme example, but if you zoom out a lot, the blue background is also wonky:
image

Maybe we need to take the flex off of it and makes sure it has a fixed height that doesn't shrink or something.

**Comment by shawnborton** [comment]
For the icon:
image


I think we should consider making that a different color. Either the same color as text in dark mode, or maybe we make it ivory, or ivory with a slightly about of transparency. I think my vote would be to make the icons ivory @ 80% opacity, and the text ("Preferences") in ivory. This way it will play nicely across both themes.

**Comment by roryabraham** [comment]
@shawnborton thanks for pointing that out. I'll work more on the layout issues. A couple clarifying questions about icon color:

- Are you referring only to the icon in this header, or the global icon color for icons everywhere?
- A best practice that we've established in many places across the app is that pressable things (such as the back button) should have three different colors depending on their state (default, hovered, pressed). If the default color is ivory w/ 80% opacity, what should be the hover color and the pressed color?

Thanks!

**Comment by roryabraham** [comment]
> Just icons in any colorful/styled empty state header

Gotcha, thanks 👍🏼

> Do our current back buttons have a hover or press style?

Looks like you're right. NAB then, but if you do want to play with it and come up with some values you like it would be an easy improvement to throw in here. Otherwise I'll just use ivory w/ opacity 80%

**Comment by shawnborton** [comment]
Let's start with ivory @ 80% and adjust from there. We might need to make adjustments after seeing all of the other background colors, too. 

What are the next steps for getting these empty states in other places?

**Comment by roryabraham** [comment]
Well after this is merged then several other pages (security, payments, help, workspaces) should be pretty trivial.

The ones that will be more involved are the ones where the header includes more than just the animation, such as the initial settings page and the About page.

**Comment by roryabraham** [comment]
I think I fixed the layout issues @shawnborton pointed out, but as of now there is a bug here with the wrong color of overscroll on iOS Safari:

https://github.com/Expensify/App/assets/47436092/f96c01cb-2d78-4929-bb97-0f6ea276bd97

**Comment by roryabraham** [comment]
Switched icon and title color to ivory:

image

**Comment by shawnborton** [comment]
Nice! Can you try it at reduced opacity, like 80%?

**Comment by shawnborton** [comment]
Nice, the height issue seems to be fixed.

**Comment by roryabraham** [comment]
Ah, missed that detail. Fixed now:

image

**Comment by roryabraham** [comment]
@allroundexperts I think this is ready for more reviews when you get a chance

File: src/hooks/useEnvironment.js
Comment (roryabraham) [reply src/hooks/useEnvironment.js:19]: Good idea. I was thinking this would be very quick but on Android it actually makes a call to the GitHub API so yeah we should improve this.

Comment (roryabraham) [reply src/hooks/useEnvironment.js]: Updated to use context

File: src/components/IllustratedHeaderPageLayout.js
Comment (roryabraham) [reply src/components/IllustratedHeaderPageLayout.js:43]: Looks like this was a remnant from another approach I was attempting and isn't needed anymore. Removed.

File: src/hooks/useDelayToggleButtonState.js
Comment (roryabraham) [reply src/hooks/useDelayToggleButtonState.js]: FWIW this was just me converting an existing HOC to a hook. Good catch on the missing `useEffect` dependency though.

Comment (roryabraham) [reply src/hooks/useDelayToggleButtonState.js]: Ended up scope-creeping a bit to clean this up because I found it confusing. I mostly followed your suggestion

File: src/components/Pressable/PressableWithDelayToggle.js
Comment (roryabraham) [inline src/components/Pressable/PressableWithDelayToggle.js:12]: No localization props were used here, so this wasn't needed anymore

File: src/components/withDelayToggleButtonState.js
Comment (roryabraham) [inline src/components/withDelayToggleButtonState.js:35]: `resetAfterDelay` was always, true, so I removed this

File: src/components/withEnvironment.js
Comment (roryabraham) [inline src/components/withEnvironment.js:26]: Why did I memoize this you ask?

Because if you pass an inline object directly to the context provider like so:

```
<EnvironmentContext.Provider value={{environment, environmentURL}}></EnvironmentContext.Provider>
```

Then the object passed to value `{environment, environmentURL}` is not referentially stable and all the context consumers will be updated on every re-render of the provider, even if the values for `environment` and `environmentURL` don't change.

To fix that, wrap the context value in `useMemo`

File: src/pages/home/report/ContextMenu/BaseReportActionContextMenu.js
Comment (roryabraham) [inline src/pages/home/report/ContextMenu/BaseReportActionContextMenu.js:123]: This was always true, so I removed the `autoReset` prop entirely

**Comment by roryabraham** [comment]
Thanks for the feedback @allroundexperts, made some improvements + cleanup here. Ready for another round of reviews 🙇🏼

**Comment by roryabraham** [comment]
Resolved conflicts, let's get this over the finish line 😃

**Comment by roryabraham** [comment]
@allroundexperts all fixed now

**Review by allroundexperts — APPROVED** [review]
Solid work @roryabraham!

I have left some minor comments. However, you can merge this if you feel like those are not urgent enough to be addressed.

**Review by MonilBhavsar — APPROVED** [review]
Looks good to me.

File: src/components/withEnvironment.js
Comment (roryabraham) [reply src/components/withEnvironment.js:1]: There are several counter-examples, where the context is created in the HOC, such as in `withLocalize`, `withCurrentReportID`, `withKeyboardState`, `withWindowDimensions`, so I'd say we're not consistent yet.

Perhaps in a follow-up we can standardize on moving all contexts to `src/contexts` and having a file for each 👍🏼

**Comment by roryabraham** [comment]
TIL about `git mv`, thanks @MonilBhavsar!

and thanks for the comments @allroundexperts. I think there's a number of good thoughts there but it's more important to get this merged so I can implement the other pages soon here.

File: src/components/HeaderWithBackButton/index.js
Comment (mountiny) [inline src/components/HeaderWithBackButton/index.js:97]: This refactor has caused a deploy blocker https://github.com/Expensify/App/pull/23129 as we have forgot to pass the children which was a change made in the deleted file but probably did not cause morege conflicts or it was missed when resolved.

Comment (roryabraham) [reply src/components/HeaderWithBackButton/index.js:97]: Yes, my bad I missed that change when resolving conflicts

File: src/components/Pressable/PressableWithDelayToggle.js
Comment (roryabraham) [reply src/components/Pressable/PressableWithDelayToggle.js:16]: My reasoning was that since it's always passed by `forwardRef`, then it would make sense to have it be required. However I see now that it will be null/undefined if the parent doesn't pass a ref.

**Comment by sobitneupane** [comment]
[This minor issue](https://github.com/Expensify/App/issues/26425) was probably missed in the PR.

Issue: Preferences - Shows blue color in Bottom Safearea when going offline

Steps to reproduce:

1. Go offline in IOS/native
2. Go to settings -> Go to preferences
3. See bottom Safearea

**Pull request merged**
